File: webclient/src/domain/frames/frameNaming.ts
/**
 * 프레임 이름 규칙·사본 이름 생성 — Windows `Frames/FrameNaming.cs` 이식 (analysis/13 §6.2·§6.4)
 */

/** 사본 접미 토큰. 파일명 규약상 `_`를 쓰지 않는다(공용/개인 구분자이므로). */
export const COPY_SUFFIX = "사본";

/** 이름이 비어 있을 때 쓰는 기본 base(결과는 "새 프레임 사본"). */
export const DEFAULT_BASE_NAME = "새 프레임";

/** 충돌 회피 번호 상한. 초과 시 8자 난수 접미로 폴백한다. */
export const MAX_COPY_INDEX = 99;

export const MAX_FRAME_NAME_LENGTH = 100;

/** `"{X} 사본"` / `"{X} 사본 N"`(N = 1~2자리) 접미 패턴. */
const COPY_SUFFIX_PATTERN = /^(.*?)\s*사본(?:\s+(\d{1,2}))?$/;

/**
 * Windows `Path.GetInvalidFileNameChars()` 대응 — `< > : " / \ | ? *` + 제어문자(0x00~0x1F).
 * 프레임을 Windows `Frame\`으로 내보낼 수 있어야 하므로(WD4) 웹에서도 같은 문자를 거부한다.
 * 공백·하이픈은 **금지문자가 아니다**(사본 이름 "새 프레임 사본"이 공백을 포함한다).
 */
// eslint-disable-next-line no-control-regex
const INVALID_FILENAME_CHARS = /[<>:"/\\|?*\u0000-\u001f]/;

/**
 * 로컬 파일명으로 쓸 수 있는 이름인지 — **빈 값·공백만이 아니고 파일시스템 금지문자가 없다**.
 * Windows `FrameNaming.IsFileNameSafe`(= `LocalFrameStore.EnsureFileNameSafe`)와 **동일 판정**이다.
 *
 * ⚠️ **길이를 보지 않는다.** `validateFrameName`은 100자 제한을 함께 보므로 **축이 다르다**
 *    (01 §2 매핑표 주석 · 03 §11.3 ⑤⑥). 저장 전 선검증에 길이 검사가 섞이면, 파워 신규 생성이
 *    "서버 insert는 성공했는데 로컬 저장은 이름 때문에 실패"하는 반쪽 상태를 못 막는 대신
 *    엉뚱한 이유로 정상 이름을 거부한다.
 *
 * ⚠️ 검사 대상은 **원문**이다(trim 결과가 아니다). 실제 파일명이 되는 문자열이 그대로 이 값이기 때문이며,
 *    Windows도 `IndexOfAny`를 원문에 건다. 그래서 `"이름\n"`은 여기서 **거부**되지만
 *    `validateFrameName`은 trim 후 판정하므로 **통과**한다 — 의도된 축 차이다.
 */
export function isFileNameSafe(name: string | null | undefined): boolean {
  if (name === null || name === undefined) return false;
  if (name.trim().length === 0) return false;
  return !INVALID_FILENAME_CHARS.test(name);
}

/**
 * `"{X} 사본"` / `"{X} 사본 N"` 접미를 제거해 원형 이름을 얻는다(접미가 없으면 원문 그대로).
 * 접미를 떼면 이름이 비게 되는 경우(예: `"사본"`)도 원문을 그대로 반환한다 — 빈 이름을 만들지 않는다.
 */
export function stripCopySuffix(name: string): string {
  if (name.trim().length === 0) return name;

  const match = COPY_SUFFIX_PATTERN.exec(name.trim());
  if (match === null) return name;

  const stripped = (match[1] ?? "").replace(/\s+$/, "");
  return stripped.trim().length === 0 ? name : stripped;
}

/**
 * `baseName` 기준으로 `existingNames`와 충돌하지 않는 사본 이름을 만든다.
 * `"{base} 사본"` → 충돌 시 `"{base} 사본 2"` … 99까지 → 그 뒤 `"{base} 사본 {8자 난수}"`.
 * 이미 사본 형태면 원형으로 되돌려 무한 누적을 막는다. 비교는 정확 일치(Ordinal).
 *
 * @param uniqueSuffix 99까지 모두 충돌할 때 쓸 8자 접미 생성기. 도메인은 난수를 직접 만들지 않는다
 *                     (어댑터가 `crypto`로 주입 — 01 §8). **항상 이름을 돌려준다**(저장을 막지 않는다).
 */
export function nextCopyName(
  baseName: string | null | undefined,
  existingNames: Iterable<string>,
  uniqueSuffix: () => string,
): string {
  let root = stripCopySuffix(baseName ?? "");
  if (root.trim().length === 0) root = DEFAULT_BASE_NAME;

  const taken = new Set<string>();
  for (const name of existingNames) {
    if (name.length > 0) taken.add(name);
  }

  const first = `${root} ${COPY_SUFFIX}`;
  if (!taken.has(first)) return first;

  for (let n = 2; n <= MAX_COPY_INDEX; n++) {
    const candidate = `${root} ${COPY_SUFFIX} ${n}`;
    if (!taken.has(candidate)) return candidate;
  }

  return `${root} ${COPY_SUFFIX} ${uniqueSuffix()}`;
}

export type FrameNameRejection = "empty" | "too-long" | "invalid-chars" | "underscore";

export interface FrameNameValidation {
  readonly ok: boolean;
  readonly reason?: FrameNameRejection;
}

/**
 * 로컬 저장용 이름 검증: 1~100자 + 파일시스템 금지문자 없음(치환하지 않고 거부).
 *
 * ⚠️ **`_`를 여기서 거부하지 않는다.** 로컬 개인 저장은 서버를 거치지 않으므로 `_`가 계약 위반이
 *    아니다. 다만 **공용** 스코프에서는 파일명 규약과 충돌하므로 `underscoreWarning`으로 안내한다
 *    (Windows `FrameEditorViewModel`과 동일한 비차단 경고).
 *    **서버에 등록하는 경로는 `validateFrameNameForServer`를 쓴다** — 서버가 400으로 거부한다.
 *
 * ⚠️ 저장 **직전** 선검증은 길이를 보지 않는 `isFileNameSafe`를 쓴다. 이름 판정은 **세 축**이다:
 *    서버 등록 = `validateFrameNameForServer` / 로컬 저장 = `validateFrameName` + `underscoreWarning` /
 *    저장 전 선검증 = `isFileNameSafe`.
 */
export function validateFrameName(name: string): FrameNameValidation {
  const trimmed = name.trim();
  if (trimmed.length === 0) return { ok: false, reason: "empty" };
  if (trimmed.length > MAX_FRAME_NAME_LENGTH) return { ok: false, reason: "too-long" };
  // 금지문자 판정은 `isFileNameSafe`가 유일한 원본이다(축 분리 후에도 두 곳에서 정규식을 쓰지 않는다).
  // **`trimmed`를 넘기는 것이 계약**이다 — 원문을 넘기면 `"이름\n"`의 판정이 바뀌어 기존 결과가 회귀한다.
  if (!isFileNameSafe(trimmed)) return { ok: false, reason: "invalid-chars" };
  return { ok: true };
}

/**
 * 서버 등록용 이름 검증 — `POST /frames`에 보낼 이름(power 신규 공용 프레임)에 쓴다.
 *
 * 로컬 규칙 + **`_` 금지**(M15). 서버 `domain/validation.ts validateFrameName`이 `_`를 400으로
 * 거부하므로, 보내기 전에 같은 판정을 해 사용자에게 즉시 사유를 알린다(성공 오인·왕복 낭비 방지).
 * 두 판정이 어긋나면 서버가 진실원이다.
 */
export function validateFrameNameForServer(name: string): FrameNameValidation {
  const local = validateFrameName(name);
  if (!local.ok) return local;
  if (name.trim().includes("_")) return { ok: false, reason: "underscore" };
  return { ok: true };
}

/** 프레임 저장 스코프. 공용 = 접두 없는 파일명, 개인 = `{계정}_{이름}`. */
export type FrameSaveScope = "public" | "personal";

/**
 * 공용 스코프 저장 시 이름에 `_`가 있으면 **비차단 경고** 대상이다.
 * 공용 목록은 파일명에 `_`가 없는 것만 집계하므로(개인 접두 구분자) 목록에서 보이지 않을 수 있다.
 * 개인 스코프는 `{계정}_` 접두가 붙는 것이 정상이라 경고하지 않는다.
 */
export function underscoreWarning(name: string, scope: FrameSaveScope): boolean {
  return scope === "public" && name.includes("_");
}

File: src/MCPhoto.Core/Frames/FrameNaming.cs
using System.IO;
using System.Text.RegularExpressions;

namespace MCPhoto.Core.Frames;

/// <summary>
/// 프레임 사본 이름 생성(순수). 서버 프레임을 불러와 새로 만들 때 원본과 이름을 구분해
/// 원본 파일을 보존하고 FrameCatalogService의 이름 기준 dedup(재다운로드 방지)을 유지한다. (it15 F1-D4)
/// </summary>
public static partial class FrameNaming
{
    /// <summary>사본 접미 기본 토큰. 파일명 규약상 '_'는 쓰지 않는다(LocalFrameStore 공용/user 구분자).</summary>
    public const string CopySuffix = "사본";

    /// <summary>이름이 비어 있을 때 사용하는 기본 base(결과는 "새 프레임 사본").</summary>
    public const string DefaultBaseName = "새 프레임";

    /// <summary>충돌 회피 번호 상한. 초과 시 GUID 8자 접미로 폴백한다.</summary>
    private const int MaxCopyIndex = 99;

    /// <summary>"{X} 사본" / "{X} 사본 N"(N=1~2자리) 접미 파싱. '_'를 새로 도입하지 않는다(§1.5 함정).</summary>
    [GeneratedRegex(@"^(?<base>.*?)\s*사본(\s+(?<n>\d{1,2}))?$", RegexOptions.CultureInvariant)]
    private static partial Regex CopySuffixPattern();

    /// <summary>
    /// baseName 기준으로 existingNames와 충돌하지 않는 사본 이름을 만든다.
    /// "{base} 사본" → 충돌 시 "{base} 사본 2", "{base} 사본 3" … 99까지.
    /// baseName이 이미 "{X} 사본" / "{X} 사본 N" 형태면 X를 base로 되돌려 무한 누적을 막는다.
    /// 99까지 모두 충돌하면 "{base} 사본 {8자리 GUID}"를 반환한다(항상 이름을 돌려준다).
    /// 비교는 <see cref="StringComparer.Ordinal"/>(LocalFrameStore 파일명 규약과 동일).
    /// </summary>
    /// <param name="baseName">원본 이름. null·공백이면 <see cref="DefaultBaseName"/>을 base로 사용.</param>
    /// <param name="existingNames">같은 저장 스코프의 기존 이름들(공용=PublicFrameNames, 개인=LoadUser).</param>
    public static string NextCopyName(string? baseName, IEnumerable<string> existingNames)
    {
        var root = StripCopySuffix(baseName ?? string.Empty);
        if (string.IsNullOrWhiteSpace(root)) root = DefaultBaseName;

        var taken = new HashSet<string>(
            existingNames.Where(n => !string.IsNullOrEmpty(n)), StringComparer.Ordinal);

        var first = $"{root} {CopySuffix}";
        if (!taken.Contains(first)) return first;

        for (int n = 2; n <= MaxCopyIndex; n++)
        {
            var candidate = $"{root} {CopySuffix} {n}";
            if (!taken.Contains(candidate)) return candidate;
        }

        // 1~99 전부 충돌: 예외 대신 항상 이름을 돌려준다(저장을 막지 않는다).
        var unique = Guid.NewGuid().ToString("N")[..8];
        return $"{root} {CopySuffix} {unique}";
    }

    /// <summary>
    /// 로컬 파일명으로 쓸 수 있는 이름인지(빈 값·공백 아님 + 파일시스템 금지문자 없음). 순수 함수.
    /// <c>LocalFrameStore.EnsureFileNameSafe</c>의 판정과 **동일**하다 — 저장 전 선검증에 쓰기 위해 추출했다.
    /// 파워 신규 생성은 서버 insert를 먼저 하고 로컬 저장을 나중에 하므로, 이 선검증 없이는 잘못된 이름에서
    /// "서버에는 등록됐지만 로컬에는 없는" 반쪽 상태가 만들어진다.
    /// </summary>
    public static bool IsFileNameSafe(string? name)
        => !string.IsNullOrWhiteSpace(name)
           && name!.IndexOfAny(Path.GetInvalidFileNameChars()) < 0;

    /// <summary>
    /// 저장 가능한 이름인가 — <b>본인에게 보이는 프레임과 이름이 겹치지 않아야 한다</b>(설계 D-17).
    /// <para>
    /// 판정 집합 = <b>공용 프레임 이름 ∪ 본인 개인 프레임 이름</b>. 다른 계정의 개인 프레임과는
    /// 겹쳐도 된다 — 저장 폴더가 계정별로 나뉘어 있고 서로 보이지도 않는다.
    /// </para>
    /// <para>
    /// 비교는 <b>대소문자 무시</b>다. Windows 파일시스템이 대소문자를 구분하지 않으므로
    /// <c>"Abc"</c>와 <c>"abc"</c>를 허용하면 실제 파일이 서로 덮어써진다.
    /// </para>
    /// ⚠️ 이 검증은 <b>즉시 피드백</b>용이다. PC 두 대에서 동시에 같은 이름을 만드는 경우는 막지 못하므로
    /// <b>서버가 계정 내 이름 중복을 최종 거부</b>해야 한다(설계 S8).
    /// </summary>
    /// <param name="name">저장하려는 이름.</param>
    /// <param name="visibleNames">현재 사용자에게 보이는 프레임 이름 전부(공용 + 본인 개인).</param>
    public static bool IsNameAvailable(string? name, IEnumerable<string>? visibleNames)
    {
        var candidate = (name ?? string.Empty).Trim();
        if (candidate.Length == 0) return false;
        if (visibleNames is null) return true;

        foreach (var taken in visibleNames)
        {
            if (string.IsNullOrWhiteSpace(taken)) continue;
            if (string.Equals(taken.Trim(), candidate, StringComparison.OrdinalIgnoreCase))
                return false;
        }
        return true;
    }

    /// <summary>
    /// "{X} 사본" / "{X} 사본 N" 접미를 제거해 원형 이름을 얻는다(접미가 없으면 원문 그대로).
    /// 접미를 떼면 이름이 비게 되는 경우(예: "사본")도 원문을 그대로 반환한다 — 빈 이름을 만들지 않는다.
    /// </summary>
    public static string StripCopySuffix(string name)
    {
        if (string.IsNullOrWhiteSpace(name)) return name;

        var m = CopySuffixPattern().Match(name.Trim());
        if (!m.Success) return name;

        var stripped = m.Groups["base"].Value.TrimEnd();
        return string.IsNullOrWhiteSpace(stripped) ? name : stripped;
    }
}
